### PR TITLE
Feat 451 add commit args

### PR DESCRIPTION
**Pull request by SCrocky**
## Description
Added extra argument mechanic for `cz commit` command so that all arguments after a `--` will be treated as extra args.
Example: You can now sign commits with `cz c -- -S`

## Checklist

- [ ] Add test cases to all the changes you introduce
- [ ] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [ ] Test the changes on the local machine manually
- [ ] Update the documentation for the changes

## Additional context
This PR should close [issue 451](https://github.com/commitizen-tools/commitizen/issues/451)

### CODE REVIEW

**Review by Lee-W — COMMENTED** [review]
Sorry for late reviewing. The overall idea looks good to me! But we'll need doc and unit test for this one. After this is merged, I think we can have one branch removing the original signoff and send that PR to `v3` branch

**Comment by SCrocky** [comment]
> Sorry for late reviewing. The overall idea looks good to me! But we'll need doc and unit test for this one. After this is merged, I think we can have one branch removing the original signoff and send that PR to `v3` branch

Sounds good, I'll get on the tests next.

**Comment by woile** [comment]
Any update on this? We'd appreciate having the tests and rebasing 🙏🏻

**Comment by SCrocky** [comment]
I have some time this weekend. I'll try to finish it then.

**Comment by SCrocky** [comment]
Done reading, added both tests and doc.
@woile @Lee-W

File: commitizen/commands/commit.py
Comment (Lee-W) [inline commitizen/commands/commit.py:94]: @woile Do you think we should add a list somewhere so we won't forget to deprecate stuff in the next release?

Comment (Lee-W) [reply commitizen/commands/commit.py:94]: @SCrocky In this implementation, if we're using the original `signoff` feature, we won't be able to use `extra_cli_args.`. Should we implement it as something like the following? (the warning is still needed)

```python
extra_cli_args = ""

if signoff:
    extra_cli_args = "-s "
```

File: commitizen/commands/commit.py
Comment (Lee-W) [inline commitizen/commands/commit.py:97]: Do we need this if-else block? Could we handle the case in `git.commit()` when `extra_cli_args` is not provided?

File: docs/commit.md
Comment (Lee-W) [inline docs/commit.md:24]:
```suggestion

```sh
cz commit -- <git-cli-args>

# e.g., cz commit --dry-run -- -a -S
```

```

File: commitizen/cli.py
Comment (Lee-W) [inline commitizen/cli.py:437]: We might need to handle the case that the user use `--` without actually providing an argument. e.g., `cz commit --`

**Review by Lee-W — COMMENTED** [review]
I am thrilled about this wonderful feature! I was able to spare some time to examine it. I have just provided a few comments and suggestions.

File: docs/commit.md
Comment (SCrocky) [reply docs/commit.md:24]: Can we keep `commitizen-args` ? It provides clarity IMO.

```sh
cz commit <commitizen-args> -- <git-cli-args>

# e.g., cz commit --dry-run -- -a -S
```

File: commitizen/commands/commit.py
Comment (SCrocky) [reply commitizen/commands/commit.py:94]: Ah, I see it. Thanks for that.

Would this work for you?

```python
        if signoff:
            out.warn("signoff mechanic is deprecated, please use `cz commit -- -s` instead.")
            extra_args = "-s " + self.arguments.get("extra_cli_args")
        else:
            extra_args = self.arguments.get("extra_cli_args")

        c = git.commit(m, extra_args=extra_args)
```

File: commitizen/cli.py
Comment (SCrocky) [reply commitizen/cli.py:437]: Sure, how should we deal with it?

Warning or error?

File: docs/commit.md
Comment (Lee-W) [reply docs/commit.md:24]: Sure 🙂 Looks better 👍

File: commitizen/cli.py
Comment (Lee-W) [reply commitizen/cli.py:437]: Warning should be enough in this case 👍

File: commitizen/commands/commit.py
Comment (Lee-W) [reply commitizen/commands/commit.py:94]: I might try something like

```python
        extra_args = self.arguments.get("extra_cli_args", "")
        if signoff:
            out.warn("signoff mechanic is deprecated, please use `cz commit -- -s` instead.")
            extra_args += " -s"

        c = git.commit(m, extra_args=extra_args)
```

Comment (Lee-W) [reply commitizen/commands/commit.py:94]: But one thing I'm thinking is whether we should handle `cz commit -s -- -s`

Comment (SCrocky) [reply commitizen/commands/commit.py:94]: I tested it and doesn't seem to be an issue with the git command:
```sh
git commit -s -s -m "test commit" 
[master (root-commit) b864659] test commit
 1 file changed, 0 insertions(+), 0 deletions(-)
```

But it is a bit messy. Warning or error?

Comment (Lee-W) [reply commitizen/commands/commit.py:94]: If it doesn't cause an error, I'm good with either silently passing or warning.

**Comment by mesaglio** [comment]
Is this wait somthing? I think `-S` [flag](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--Sltkeyidgt) its good to use for security reasons.

**Comment by SCrocky** [comment]
I've fixed rebased to master, fixed all the previous issues, added/edited tests and added some documentation.

@Lee-W ready for Round 2?

File: commitizen/git.py
Comment (Lee-W) [inline commitizen/git.py:103]: Would it be possible for us to combine these 2? Not sure what's the difference between them

**Review by Lee-W — COMMENTED** [review]
@woile @noirbizarre Might need some time to take a deeper look, but I think most of the parts are good.

File: commitizen/git.py
Comment (SCrocky) [reply commitizen/git.py:103]: `extra_args` are arguments that we just want to pass to `git` transparently (like `-S`).

Having the `extra_args` argument separate helps with testing and clarity.

For example:`extra_args` has to start with "--" and contain at least one flag behind it. (e.g. `"-- -S"`). 

If this is contained within the `args` argument, then it becomes a little harder to know whether or not any extra args were passed.

> Would it be possible for us to combine these 2?

`git.commit` is only called with `extra_args` in [commands/commit.py `__call__` method](https://github.com/SCrocky/commitizen/blob/feat-451-add-commit-args/commitizen/commands/commit.py#L108) where it doesn't have any `args`.

So technically, if I'm not missing anything, I could just use the `args` argument to pass the extra_args. 

> Not sure what's the difference between them

I'll add a Docstring

Comment (Lee-W) [reply commitizen/git.py:103]: > So technically, if I'm not missing anything, I could just use the args argument to pass the extra_args.

Yep, I think that's what we should do. Thanks for clarifying!

Comment (SCrocky) [reply commitizen/git.py]: Done

**Review by Lee-W — APPROVED** [review]
Thanks, @SCrocky, for your contribution! This really helps a lot. Could you please make it ready for review?

@woile @noirbizarre I'm planning on merging it these days. Please let me know if you want to take a look

**Comment by SCrocky** [comment]
> Thanks, @SCrocky, for your contribution! This really helps a lot.

No problem! I'm really glad to have finally finished it. :pray: 

> Could you please make it ready for review?

Is this addressed to me? I'm sorry, I'm not sure what the remaining steps are. Is there anything else I need to do?

**Comment by Lee-W** [comment]
> > Thanks, @SCrocky, for your contribution! This really helps a lot.
> 
> No problem! I'm really glad to have finally finished it. 🙏
> 
> > Could you please make it ready for review?
> 
> Is this addressed to me? I'm sorry, I'm not sure what the remaining steps are. Is there anything else I need to do?

Ah, just found out I can do it as well

**Pull request merged**
